codec/src/types/primitives.rs

```diff
@@ -145,19 +201,27 @@ impl_codec_for_tuple!(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11);
 // Vec implementation
 impl<T: Codec> Codec for Vec<T> {
     #[inline]
-    fn write(&self, writer: &mut impl Writer) {
-        writer.write_vec(self);
+    fn write<B: BufMut>(&self, buf: &mut B) {
+        self.len().write(buf);
+        for item in self {
+            item.write(buf);
+        }
     }
 
     #[inline]
     fn len_encoded(&self) -> usize {
-        let len = varint::varint_size(self.len() as u64);
+        let len = varint::size(self.len() as u64);
         self.iter().map(Codec::len_encoded).sum::<usize>() + len
     }
 
     #[inline]
-    fn read(reader: &mut impl Reader) -> Result<Self, Error> {
-        reader.read_vec()
+    fn read<B: Buf>(buf: &mut B) -> Result<Self, Error> {
```

Comment:
It is very very important we retain the ability to do `read_lte`

Reply:
I'd almost go as far as saying that we shouldn't allow a `read` of vec without a max

Reply:
I agree. Due to the size of this change and the large design space, let's do this in a separate future PR.

Ticket here for discussion https://github.com/commonwarexyz/monorepo/issues/687